doc/adr/0004-global-snapshot-trigger.md

```diff
@@ -0,0 +1,23 @@
+# 4. Global snapshot trigger
+
+Date: 2022-01-05
+
+## Status
+
+Proposed
+
+## Context
+
+Global snapshot must have a deterministic trigger so then snapshots created on
```

Comment:
This is the main issue @buckysballs. We should discuss that on Friday and you can put your thoughts here anyway.

Reply:
We should use the same two triggers we have in Constellation

- Height diff
- Fallback to timed based trigger

Reply:
The issue is that those data we have on L1 but snapshot must be made on L0. It means, that either snapshot is triggered by receiving data from L1 (see 1.) or by some deterministic algorithm (see 2.).

1) L1 sends blocks to L0 and alongside sends metadata (like tips). L0 watches for these data and trigger a new snapshot on the height diff (based on received tips). There are 2 drowbacks. Firstly, there will be an additional overhead to send metadata over network, secondly L0 will depend on L1 and will not be able to create a snapshot without it.

2) We can agree on some conditions on both L0 and L1, like there are only 10 parents allowed on each height.
On L1, if there is 11th parent selected, node ignores it (or increments its height). On L0, if node received blocks and there are 10 parents at the same height, it means there will be no more blocks on that height and node can trigger a snapshot.

Reply:
Given the need to update L1 nodes with metadata from snapshots, and considering that it's pretty much as we had it in Constellation, I would implement option 1 before optimizing based on determinism.

Also, option 1 would be able to create a snapshot without any new blocks created with a timed based trigger. in this case, call it an "empty anapshot", we just need to set edge from the previous snapshot as both the tips for the new one

Reply:
Can you elaborate more on this part?
> we just need to set edge from the previous snapshot as both the tips for the new one

For now there were no tips in snapshots

Reply:
About time-based: How we can then figure out who should be the next facilitator if we do not have metadata from L1?

Reply:
So a snapshot is basically just a collection of blocks that is path connected to the collection of blocks in the proceeding snapshot. So basically, it's a linear data structure that links clusters of blocks into a topological order. If there are no blocks received and a snapshot needs to be created via timed trigger, then the proposal can be of one single block, with tips coming from the highest order (sorted by height, width depth [use Ordering]); or in the case of an 'empty' snapshot proceeding an 'empty' snapshot, just use the previous block in the 'empty' snapshot as both tips for the edge